File: packages/ink-swift/Sources/InkCore/Rng.swift
/// Deterministic RNG for sampling: sfc32 seeded via splitmix32, normal
/// deviates via Box-Muller. A port of packages/ink-graves/src/rng.ts; the
/// integer path is bit-identical (32-bit wrapping arithmetic), so a seed
/// consumes the uniform stream in the same order as the web engine.

import Foundation

public struct Rng: Sendable {
    private var a: UInt32
    private var b: UInt32
    private var c: UInt32
    private var d: UInt32
    private var spare: Double? = nil

    public init(seed: UInt32) {
        // splitmix32 to spread one 32-bit seed into four state words
        var s = seed
        func split() -> UInt32 {
            s = s &+ 0x9e37_79b9
            var z = s
            z = (z ^ (z >> 16)) &* 0x21f0_aaad
            z = (z ^ (z >> 15)) &* 0x735a_2d97
            return z ^ (z >> 15)
        }
        a = split()
        b = split()
        c = split()
        d = split()
        for _ in 0 ..< 12 { _ = uniform() }
    }

    /// Uniform in [0, 1).
    public mutating func uniform() -> Double {
        let t = a &+ b
        a = b ^ (b >> 9)
        b = c &+ (c << 3)
        c = (c << 21) | (c >> 11)
        d = d &+ 1
        let out = t &+ d
        c = c &+ out
        return Double(out) / 4294967296
    }

    /// Standard normal deviate.
    public mutating func normal() -> Double {
        if let value = spare {
            spare = nil
            return value
        }
        var u1 = uniform()
        while u1 == 0 { u1 = uniform() }
        let u2 = uniform()
        let radius = (-2 * Foundation.log(u1)).squareRoot()
        spare = radius * sin(2 * .pi * u2)
        return radius * cos(2 * .pi * u2)
    }

    /// Sample an index proportionally to non-negative weights.
    public mutating func categorical(_ weights: [Float]) -> Int {
        var total = 0.0
        for weight in weights { total += Double(weight) }
        var target = uniform() * total
        for (index, weight) in weights.enumerated() {
            target -= Double(weight)
            if target < 0 { return index }
        }
        return weights.count - 1
    }
}

File: packages/ink-kotlin/ink-core/src/main/kotlin/com/trylonghand/ink/core/Rng.kt
/// Deterministic RNG for sampling: sfc32 seeded via splitmix32, normal
/// deviates via Box-Muller. A port of packages/ink-graves/src/rng.ts; the
/// integer path is bit-identical (32-bit wrapping arithmetic, which Kotlin's
/// UInt operators give natively), so a seed consumes the uniform stream in
/// the same order as the web engine.

package com.trylonghand.ink.core

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.ln
import kotlin.math.sin
import kotlin.math.sqrt

public class Rng(seed: UInt) {
    private var a: UInt = 0u
    private var b: UInt = 0u
    private var c: UInt = 0u
    private var d: UInt = 0u
    private var spare: Double? = null

    init {
        // splitmix32 to spread one 32-bit seed into four state words
        var s = seed
        fun split(): UInt {
            s += 0x9e37_79b9u
            var z = s
            z = (z xor (z shr 16)) * 0x21f0_aaadu
            z = (z xor (z shr 15)) * 0x735a_2d97u
            return z xor (z shr 15)
        }
        a = split()
        b = split()
        c = split()
        d = split()
        repeat(12) { uniform() }
    }

    /** Uniform in [0, 1). */
    public fun uniform(): Double {
        val t = a + b
        a = b xor (b shr 9)
        b = c + (c shl 3)
        c = (c shl 21) or (c shr 11)
        d += 1u
        val out = t + d
        c += out
        return out.toDouble() / 4294967296.0
    }

    /** Standard normal deviate. */
    public fun normal(): Double {
        spare?.let {
            spare = null
            return it
        }
        var u1 = uniform()
        while (u1 == 0.0) u1 = uniform()
        val u2 = uniform()
        val radius = sqrt(-2 * ln(u1))
        spare = radius * sin(2 * PI * u2)
        return radius * cos(2 * PI * u2)
    }

    /** Sample an index proportionally to non-negative weights. */
    public fun categorical(weights: FloatArray): Int {
        var total = 0.0
        for (weight in weights) total += weight.toDouble()
        var target = uniform() * total
        for (index in weights.indices) {
            target -= weights[index].toDouble()
            if (target < 0) return index
        }
        return weights.size - 1
    }
}
